File: [20]validParentheses.java
/*
* Runtime: 2 ms, faster than 54.53% of Java online submissions for Valid Parentheses.
* Memory Usage: 34.3 MB, less than 100.00% of Java online submissions for Valid Parentheses.
* @dataStrucutre: stack 
*/

class Solution {
    // use HashMap class to build ST
    private HashMap<Character, Character> map = new HashMap<>();
    // trick: ST simplify code
    public boolean isValid(String s) {
        map.put(')', '(');
        map.put(']', '[');
        map.put('}', '{');
        Stack<Character> stack = new Stack<>(); // Stack class
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (this.map.containsKey(c)) {
                char top = stack.empty()? '#' : stack.pop(); // boolean ? case1 : case2
                if (top != map.get(c)) return false;
            }
            else stack.push(c);
        }
        return stack.isEmpty(); // may have extra parantheses
    }
}

System.out.println(Solution.isValid( '{[]]()]}' ));

/*
time complexity: worst O(n)
space complexity: worst O(n)
*/

File: [20]validParentheses.py
# Runtime: 24 ms, faster than 94.61% of Python3 online submissions for Valid Parentheses.
# Memory Usage: 12.8 MB, less than 100.00% of Python3 online submissions for Valid Parentheses.


class Solution:
    def isValid(self, s: str) -> bool:
        map = {')':'(', ']':'[', '}':'{'}
        stack = [];
        for char in s:
            if char in map:
                top = stack.pop() if stack else '#'
                if top != map[char]:
                    return False
            else:
                stack.append(char)
        return not stack
